### ISSUE TITLE
Idle pool can run out of space

### ISSUE THREAD

**Issue by nedbass** [zfsonlinux/zfs#2802]
Issue originally seen on FreeBSD and reported on the illumos-zfs mailing list:

http://www.listbox.com/member/archive/182191/2014/10/sort/thread/page/1/entry/9:21/20141015004531:13B211A0-5426-11E4-B22C-D98E10DF187B/

I tested the send stream posted in that thread and verified the bug affects the Linux port as well.  Here is a patch proposed in the mailing list thread that @ahrens thought looks correct. 

```diff
--- sys/cddl/contrib/opensolaris/uts/common/fs/zfs/dsl_scan.c   (revision 273115)
+++ sys/cddl/contrib/opensolaris/uts/common/fs/zfs/dsl_scan.c   (working copy)
@@ -1476,12 +1476,6 @@ dsl_scan_sync(dsl_pool_t *dp, dmu_tx_t *tx)
                           "traverse_dataset_destroyed()", err);
               }

-               /*
-                * If we didn't make progress, mark the async destroy as
-                * stalled, so that we will not initiate a spa_sync() on
-                * its behalf.
-                */
-               scn->scn_async_stalled = (scn->scn_visited_this_txg == 0);

               if (bptree_is_empty(dp->dp_meta_objset, dp->dp_bptree_obj)) {
                       /* finished; deactivate async destroy feature */
@@ -1495,6 +1489,13 @@ dsl_scan_sync(dsl_pool_t *dp, dmu_tx_t *tx)
                           dp->dp_bptree_obj, tx));
                       dp->dp_bptree_obj = 0;
                       scn->scn_async_destroying = B_FALSE;
+               } else {
+                       /*
+                        * If we didn't make progress, mark the async destroy as
+                        * stalled, so that we will not initiate a spa_sync() on
+                        * its behalf.
+                        */
+                       scn->scn_async_stalled = (scn->scn_visited_this_txg == 0);
               }
       }
       if (scn->scn_visited_this_txg) {

```

**Comment by behlendorf** [comment]
We're currently carrying the FreeBSD version of this fix.  For the next release we should replace it with the finalized version from illumos.  illumos/illumos-gate@231aab857f14a3e7a0ed5f2879399c3cd6ae92ea

**Comment by behlendorf** [comment]
The rest of the upstream changes were merged as 905edb405da278ccb019c656408f82796a344510. We're back in sync.  Closing.
